nixos/modules/config/stub-ld.nix

```
@@ -0,0 +1,76 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) last splitString mkOption types mdDoc mkIf optionals;
+
+  message = ''
+    You seem to be trying to run a dynamically linked executable that
+    wasn't made for NixOS. This is possible, but not without extra steps,
+    because NixOS, by design, does not have a global library path.
+
+    There are several solutions available:
+
+    1. If the program is packaged in nixpkgs, use that version.  You
+       can search available packages at https://search.nixos.org/packages
+
+    2. Write a nix expression for the program to package it
+       yourself. There are multiple approaches to this:
+
+       a. Build from source. This is ideal.
+
+       b. Fix the program's ELF metadata to hardcode paths to libraries
+          using autoPatchelfHook. Use this if building from source isn't
+          feasible.
+
+       c. Wrap the program to run in a normal-seeming container using
+          buildFHSEnv. This is undesirable, but sometimes necessary, for
+          example, if the package downloads and runs further executables.
+
+    3. If you expect to be doing this sort of thing a lot, you can set up
+       nix-ld using programs.nix-ld.enable. Make sure to add any needed
+       dependencies to programs.nix-ld.libraries.
+
+    4. Try running your program with steam-run, which uses the same
+       buildFHSEnv container as the steam package. This is quick and easy,
+       and usually works, but isn't a particularly good long-term solution.
+  '';
+
+  rule-for = pkgs: let
+    stub-ld = pkgs.pkgsStatic.runCommandCC "stub-ld" {
+      nativeBuildInputs = [ pkgs.unixtools.xxd ];
+      inherit message;
+    } ''
+      printf "%s" "$message" | xxd -i -n message >main.c
+      cat <<EOF >>main.c
+      #include <stdio.h>
+      int main() {
+        fwrite(message, sizeof(unsigned char), message_len, stderr);
+        return 1;
+      }
+      EOF
+      $CC main.c -o $out
+    '';
+
+    inherit (pkgs.stdenv.hostPlatform) libDir;
+    ldsoBasename = last (splitString "/" pkgs.stdenv.cc.bintools.dynamicLinker);
+  in
+    "L+ /${libDir}/${ldsoBasename} - - - - ${stub-ld}";
+in {
+  options.environment.stub-ld = mkOption {
+    type = types.bool;
+    default = true;
```

I would give this a month for people to test:

```suggestion
    default = false;
```

It's a simple program but it effects every NixOS user, so we got to be a bit careful.

---

I can see where you're coming from but I think the only good way to actually test this is to unleash it onto users' systems.

This is also a module that is mostly useful to those with little NixOS experience, so it should absolutely be enabled by default at some point.

---

I agree with Atemu here. We're not going to get meaningful testing out of starting with it disabled. The people who need it definitely don't know to enable it, and the rest are unlikely to enable it, or even notice its existence.